src/Data/Vec/Properties.agda

```diff
@@ -691,8 +696,209 @@ foldr-fusion : ∀ {A : Set a}
+                  (∀ {c} {C} {g : ∀ {n} → C n → A → C (suc n)} e →
+                   ∀ {n} x →
+                   (h {c} C g e {suc n}) ∘ (x ∷_) ≗ h (C ∘ suc) (λ {n} → g {suc n}) (g e x)) →
+                  ∀ {n} → h B f e ≗ foldl B {n} f e
+foldl-universal B f {e} h base step []       = base e
+foldl-universal B f {e} h base step (x ∷ xs) = begin
+  h B f e (x ∷ xs)
+    ≡⟨ step e x xs ⟩
+  h (B ∘ suc) (λ {n} → f {suc n}) (f e x) xs
+    ≡⟨ foldl-universal (B ∘ suc) (λ {n} → f {suc n}) {f e x} h base step xs ⟩
+  foldl (B ∘ suc) (λ {n} → f {suc n}) (f e x) xs
+    ≡⟨⟩
+  foldl B f e (x ∷ xs)
+    ∎
+  where open P.≡-Reasoning
+
+foldl-[] : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → B n → A → B (suc n)) {e} → foldl B f e [] ≡ e
+foldl-[] B f = refl
+
+foldl-fusion : ∀ {A : Set a}
+               {B : ℕ → Set b} {C : ℕ → Set c}
+               (h : ∀ {n} → B n → C n) →
+               {f : ∀ {n} → B n → A → B (suc n)} (d : B zero) →
+               {g : ∀ {n} → C n → A → C (suc n)} (e : C zero) →
+               (h {zero} d ≡ e) →
+               (∀ {n} b x → h (f {n} b x) ≡ g (h b) x) →
+               ∀ {n} → h ∘ foldl B {n} f d ≗ foldl C g e
+foldl-fusion h {f} d {g} e base fuse []       = base
+foldl-fusion h {f} d {g} e base fuse (x ∷ xs) =
+  foldl-fusion (λ {n} → h {suc n}) (f d x) (g e x) eq fuse xs
+  where
+    open P.≡-Reasoning
+    eq : h (f d x) ≡ g e x
+    eq = begin
+       h (f d x) ≡⟨ fuse d x ⟩
+       g (h d) x ≡⟨ P.cong (λ e → g e x) base ⟩
+       g e x     ∎
+
+------------------------------------------------------------------------
+-- foldr, foldl and _-∷ʳ_
+
+foldr-∷ʳ : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → A → B n → B (suc n)) {n} e y ys →
+           foldr B {suc n} f e (ys ∷ʳ y) ≡ foldr (B ∘ suc) {n} f (f y e) ys
+foldr-∷ʳ B f e y [] = refl
+foldr-∷ʳ B f e y (x ∷ xs) = P.cong (f x) (foldr-∷ʳ B f e y xs)
+
+foldl-∷ʳ : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → B n → A → B (suc n)) {n} e y ys →
+           foldl B {suc n} f e (ys ∷ʳ y) ≡ f (foldl B {n} f e ys) y
+foldl-∷ʳ B f e y []       = refl
+foldl-∷ʳ B f e y (x ∷ xs) = foldl-∷ʳ (B ∘ suc) f (f e x) y xs
+
+-- A foldr after a reverse is a foldl.
+
+foldr-ʳ++ : ∀ (B : ℕ → Set b) (f : ∀ {n} → A → B n → B (suc n))
+           {m} {n} b (xs : Vec A m) {ys : Vec A n} →
+           foldr B f b (xs ʳ++ ys)
+           ≡
+           foldl (B ∘ (_+ n)) ((λ {m} → flip (f {m + n}))) (foldr B f b ys) xs
+foldr-ʳ++ {A = A} B f {m} {n} b xs {ys} =
+  foldl-fusion (foldr B f b) ys (foldr B f b ys) refl (λ b x → refl) xs
+
+------------------------------------------------------------------------
+-- _∷ʳ_
+
+module _ {x y : A} where
+
+  ∷ʳ-injective : ∀ {n} (xs ys : Vec A n) → xs ∷ʳ x ≡ ys ∷ʳ y → xs ≡ ys × x ≡ y
+  ∷ʳ-injective []          []          refl = (refl , refl)
+  ∷ʳ-injective (x ∷ xs)    (y  ∷ ys)   eq   with ∷-injective eq
+  ... | refl , eq′ = Prod.map (P.cong (x ∷_)) id (∷ʳ-injective xs ys eq′)
```

Comment:
Can use `Prod.map_1` here (sorry for the latex notation!)

Reply:
Nice tip! Thanks.